### DISCUSSION TITLE
引用

### DISCUSSION THREAD

**Discussion by jimmyd95** [TheNetAdmin/zjuthesis#219] [General]
我正在寫英語的論文，路徑沒有問題，但是不知道爲什麽沒辦法把參考文獻寫出來。求教。

可以告訴我一下body/ref.bib是在哪裏被用的麽？可能是之前修改太多東西了導致一些頁面沒有原來的參考文獻了？

我在graduate-eng/post.tex裏面用了

\begin{thebibliography}{100}
    \linespreadsingle{}
    \printbibliography[title={Reference}]
\end{thebibliography}

然後出來了一個單一的中文“参考文献”。但是我把undergrad和doctorate的所有文檔都刪掉了，不應該會有中文的啊？？？？

**Comment by TheNetAdmin** [comment]
按理说是这里

https://github.com/TheNetAdmin/zjuthesis/blob/41866751c14e5982498e3faccbb5d65d3347baf8/body/graduate-eng/post.tex#L4

我不知道你对模板做了哪些修改，我建议你尝试下载一个新的模板，一点一点加入你的内容，加入一点就重新编译一遍，看是哪些代码出了问题。

如果还有问题，可以开个issue，填好相关信息，我可以帮你看一下

**Reply by jimmyd95** [reply]
我就是修改了這個地方修改了老半天……還是沒辦法解決問題。明明title=Reference，但是出來的説是“參考文獻”。新年快樂，抱歉沒來得及和回復

PS. 我把所有和master’s 無關的都刪掉了，就只剩下grad, grad-eng。連phd都沒有，什麽都沒有留下（因爲我覺得沒什麽相關聯這樣容易讓我上手）

因爲我經常在不同地方寫論文，所以給每一個電腦本地編譯實在是太麻煩。有可能是因爲overleaf的問題麽？

![loomed_debug](https://user-images.githubusercontent.com/16414366/152302127-bf164a30-53f9-417d-8d0b-67060446cf9e.png)
![loomed_debug2](https://user-images.githubusercontent.com/16414366/152302256-94c187b0-4dce-4201-aa72-39c4853b74e6.png)

**Reply by TheNetAdmin** [reply]
我不建议你对 `page` 目录下的内容做过多删减，有些页面是多个学位共用的。但如果你没遇到问题那就暂时不要改回去了。

**Comment by hushidong** [comment]
修改bibname试试？

**Reply by jimmyd95** [reply]
額……github好像出現bug了，我不確定你到底發了什麽東西……一大堆的js和HTML指令

**Comment by hushidong** [comment]
你会改吧，renewcommand改一下bibname的定义。

**Reply by hushidong** [reply]
我用手机所以不方便写完整命令。

**Reply by jimmyd95** [reply]
嗯，我在post裏面嘗試一下

**Reply by jimmyd95** [reply]
不行，現在我是直接在post.tex裏面添加引用。

\cleardoublepage
\renewcommand\bibname{Reference}
\begingroup
\begin{thebibliography}{99}

\bibitem{} 神奇的引用

\end{thebibliography}
\endgroup

寫了這樣的，pdf裏面才會出現引用……我恢復你之前的那個就直接沒有反應了

**Comment by hushidong** [comment]
zjuthesis使用biblatex-gb7714-2015，使用如下三种方式设置都是可以的：
方式一：
```
\documentclass{ctexart}
\usepackage[style=gb7714-2015]{biblatex}

\begin{filecontents}{\jobname.bib}
@Article{丁文祥2000--,
  Title                    = {数字革命与竞争国际化},
  Author                   = {丁文祥},
  Date                     = {2000-11-20},
  Journaltitle             = {中国青年报},
  Number                   = {15},
  Note                     = {News}
}
\end{filecontents}

\addbibresource{\jobname.bib}



\begin{document}

\nocite{*}

\sloppy
\printbibliography[title=reference]

\end{document} 
```

结果：
![图片](https://user-images.githubusercontent.com/20421795/152348781-7c927b93-c554-4a7f-8fc1-8ab9f8c0af54.png)

方式二：
```
\documentclass{ctexart}
\usepackage[style=gb7714-2015]{biblatex}
\ctexset{
bibname=reference
}

\begin{filecontents}{\jobname.bib}
@Article{丁文祥2000--,
  Title                    = {数字革命与竞争国际化},
  Author                   = {丁文祥},
  Date                     = {2000-11-20},
  Journaltitle             = {中国青年报},
  Number                   = {15},
  Note                     = {News}
}
\end{filecontents}

\addbibresource{\jobname.bib}



\begin{document}

\nocite{*}

\sloppy
\printbibliography

\end{document} 
```
或
```
\documentclass{ctexart}
\usepackage[style=gb7714-2015]{biblatex}
\newcommand{\bibname}{reference}
\renewcommand{\refname}{reference}

\begin{filecontents}{\jobname.bib}
@Article{丁文祥2000--,
  Title                    = {数字革命与竞争国际化},
  Author                   = {丁文祥},
  Date                     = {2000-11-20},
  Journaltitle             = {中国青年报},
  Number                   = {15},
  Note                     = {News}
}
\end{filecontents}

\addbibresource{\jobname.bib}



\begin{document}

\nocite{*}

\sloppy
\printbibliography

\end{document} 
```

结果为：
![图片](https://user-images.githubusercontent.com/20421795/152348960-927ac869-e0a8-42e3-b0f4-3f5eb9a741e5.png)


方式三：
```
\documentclass{ctexart}
\usepackage[style=gb7714-2015,gbctexset=false]{biblatex}

\DefineBibliographyStrings{english}{
bibliography={reference},
references={reference}
}


\begin{filecontents}{\jobname.bib}
@Article{丁文祥2000--,
  Title                    = {数字革命与竞争国际化},
  Author                   = {丁文祥},
  Date                     = {2000-11-20},
  Journaltitle             = {中国青年报},
  Number                   = {15},
  Note                     = {News}
}
\end{filecontents}

\addbibresource{\jobname.bib}



\begin{document}

\nocite{*}

\sloppy
\printbibliography

\end{document} 
```

结果为：
![图片](https://user-images.githubusercontent.com/20421795/152349083-f5358ea5-cc10-46c5-a934-5e6da2e715a6.png)

**Reply by jimmyd95** [reply]
嗯……我寫的不一樣，因爲弄之前我就已經整理好一個引用的形式了……是直接這樣寫：

\bibitem{} Spek, A., Li, Wai Ho., \& Drummond, T. (2017). A Fast Method For Computing Principal Curvatures From Range Images. ArXiv.

之前可以直接出來，但是最近不行了……不知道爲什麽

**Reply by TheNetAdmin** [reply]
建议你改成 @hushidong 提到的格式，zjuthesis 模板也是在 `body/ref.bib` 中存放引用。不推荐使用 `\bibitem{}` 的方式。你可以先改一两个引用看是否有效，再考虑全部迁移。

如果仍有问题，请开 issue 讨论，这里得到的信息不足。

**Reply by jimmyd95** [reply]
我修改過了，并沒有任何效果。這就是爲啥我就物理性把所有引用寫在了post.tex裏面了。是真的沒轍了我才決定這麽搞的。

**Reply by TheNetAdmin** [reply]
这么讨论下去也不会有什么结果，把你的论文内容去除，只留下参考文献，然后打包发在这里，我给你看一下好了

**Reply by jimmyd95** [reply]
額……不知道爲啥放不進去，30MB的容量。我都不知道爲啥這麽大……

PS. 圖片都刪掉了還是放不進去

**Reply by TheNetAdmin** [reply]
Email 给我好了，zxwang42 [at] gmail.com

**Reply by TheNetAdmin** [reply]
我正在看你的代码，里面有很多 LaTeX 报错的地方，在我的本地编译环境里都无法编译。

我猜测可能是 Overleaf 帮你自动修正或者强行编译了pdf，所以导致一些奇怪的问题。

我还在修，等一会儿修好再看参考文献的问题。

**Reply by jimmyd95** [reply]
嗯……因爲我不需要用中文全程都是英文的，大部分報錯都是報錯沒有中文字體。至少之前報40多個錯誤裏面有30多個是中文字體所以我就沒管它。不過或許有其他的？不是很確定……

**Reply by TheNetAdmin** [reply]
我看了你的代码，你的主要问题在于，没有使用 @hushidong 提到的方式去输出参考文献。

在你的源代码里，`body/graduate-eng/post.tex` 你是这样写的

```TeX
\cleardoublepage
\renewcommand\bibname{References}
%\begingroup
%    \linespreadsingle{}
%    \printbibliography[title={References}]
%\endgroup
\renewcommand{\contentsname}{References}

\begin{thebibliography}{99}
\linespreadsingle{}

\bibitem{} Amenta, N., Bern, M., \& Kamvysselis, M. (1998). A new VORONOI-BASED surface reconstruction algorithm. Proceedings of the 25th Annual Conference on Computer Graphics and Interactive Techniques  - SIGGRAPH '98. https://doi.org/10.1145/280814.280947

...

\end{thebibliography}
```

你这里完全没用到这一行，因为你把它注释掉了

```TeX
\printbibliography[title={References}]
```

所以也就无法正确打印标题

我把你的 `body/graduate-eng/post.tex` 改回这样，**注意没有后面的参考文献条目**

```TeX
\cleardoublepage
\renewcommand\bibname{References}
\begingroup
   \linespreadsingle{}
   \printbibliography[title={References}]
\endgroup
```

然后再把 `body/ref.bib` 改成标准格式的引用

```bibtex
% use `texdoc biblatex` to get help
@www{zjuthesisrules,
    title = {浙江大学本科生毕业论文（设计）编写规则},
    author = {浙江大学本科生院},
    year = {2018},
    url = {http://bksy.zju.edu.cn/attachments/2018-01/01-1517384518-1149149.pdf},
}
@www{tikz,
    title = {tikz宏包},
    author = {Till Tantau},
    year = {2018},
    url = {https://sourceforge.net/projects/pgf/},
}
@www{zjuthesis,
    title = {浙江大学毕业设计/论文模板},
    author = {王子轩},
    year = {2019},
    url = {https://github.com/TheNetAdmin/zjuthesis},
}
@www{zjugradthesisrules,
    title = {浙江大学研究生学位论文编写规则},
    author = {浙江大学研究生院},
    year = {2008},
    url = {http://grs.zju.edu.cn/redir.php?catalog_id=10038&object_id=12877},
}
```

就可以正常显示英文参考文献标题了

![image](https://user-images.githubusercontent.com/18525442/152473433-1bf55e9a-d87a-4606-a7a4-66cd285e7d8c.png)

所以你根据我的改法，把 `body/graduate-eng/post.tex` 修复，然后在 `body/ref.bib` 里使用标准格式定义参考文献，就没有问题了。

**Reply by jimmyd95** [reply]
唉....花了我好長時間搞那個APA各式的......現在又要重新搞了。

好的謝謝！！我修改一下。

**Reply by TheNetAdmin** [reply]
APA 格式可以通过 biblatex 输出，建议阅读

https://www.overleaf.com/learn/latex/Biblatex_citation_styles

然后修改

https://github.com/TheNetAdmin/zjuthesis/blob/41866751c14e5982498e3faccbb5d65d3347baf8/config/packages.tex#L14-L18

应该将 `gb7714-2015` 替换为 `apa` 即可，但我没有测试，你可以试一下

**Reply by jimmyd95** [reply]
嘗試了一邊，apa修改的和gb7714-2015的，還是不行。目測應該不是格式的問題。overleaf裏面的末頁還是沒有顯示任何的結果。參考文獻就直接不見了，你還有哪裏修改了嗎？

**Reply by jimmyd95** [reply]
![1](https://user-images.githubusercontent.com/16414366/152479155-18b4a11d-bff6-4be1-a1a3-cc270f8eeaaa.png)
![2](https://user-images.githubusercontent.com/16414366/152479173-8f6ae350-4b60-4e9b-8d20-5297735457da.png)


目前就是一樣的排版，還是沒辦法在overleaf裏面顯示文獻。多出來的一個ReferenceHolder是爲了記錄我寫的參考文獻，純粹筆記本用處

**Reply by TheNetAdmin** [reply]
https://www.overleaf.com/7612767526tgnspmhbnsxw

我发给你的的代码可以在overleaf上用，不行你就把这个链接的overleaf复制一份，然后把你的内容再复制粘贴进来得了

我不知道你改了哪里或者用了什么选项导致了这些问题

**Reply by jimmyd95** [reply]
我對比一下

**Reply by jimmyd95** [reply]
我好像找到問題的答案了。你大部分的代碼都是在套用graduate而不是graduate-eng。所以我只要在graduate-eng裏面的content寫任何東西，都會直接省略掉參考文獻。你可以試試看。只要添加任何東西都會默認刪除參考文獻的頁面

只要把這段刪掉，就完全無法顯示引用：

本模板根据浙江大学研究生院编写的《浙江大学研究生学位论文编写规则》~\cite{zjugradthesisrules}，
在原有的 zjuthesis 模板~\cite{zjuthesis}基础上开发而来。

我不知道爲什麽……

**Reply by jimmyd95** [reply]
我不介意放這段啦……但是至少讓我弄成英文的吧？不然通篇英文的論文裏面一段中文有點突兀……

**Reply by hushidong** [reply]
换成apa，原模板是没有问题的。

```
\usepackage[
    style=apa,
    % gbpub=false,         % Uncomment if you do NOT want '[S.l. : s.n.]' in reference entries, GitHub Issue (#47)
    % gbnamefmt=lowercase, % Uncomment if you do NOT want uppercase author names in reference entries, GitHub Issue (#23)
]{biblatex}
```
```
\printbibliography[title=reference]
```

结果为：

![图片](https://user-images.githubusercontent.com/20421795/152491787-6840c026-539c-4755-a1cd-285f2233a174.png)


剩下的就是中文改英文的问题，根据需要改就行了。

**Reply by jimmyd95** [reply]
還是不行，你給我的裏面也是，只要在grad-eng裏面的content刪掉那段就完全沒有參考文獻了……

和apa修改沒有關係，你給我的overleaf都不行

Graduate-eng裏content.tex只要換一下就完全沒有參考文獻了。算了感覺弄不了我直接物理性把引用加入post.tex裏面吧，謝謝幫忙解答！！
